### PR TITLE
Preparing the API for wis2box-1.0b5 

### PR DESCRIPTION
**Pull request by maaikelimper**
New wis2box-api for wis2box-1.0b5 release:
- wis2box-synop-process: synop2bufr using API-call
- wis2box-csv-process: csv2bufr using API-call, including aws_mappings template
- added testing

### CODE REVIEW

File: docker_compose_test/test_csv_publish.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    "data": "wsi_series,wsi_issuer,wsi_issue_number,wsi_local,wmo_block_number,wmo_station_number,station_type,year,month,day,hour,minute,latitude,longitude,station_height_above_msl,barometer_height_above_msl,station_pressure,msl_pressure,geopotential_height,thermometer_height,air_temperature,dewpoint_temperature,relative_humidity,method_of_ground_state_measurement,ground_state,method_of_snow_depth_measurement,snow_depth,precipitation_intensity,anemometer_height,time_period_of_wind,wind_direction,wind_speed,maximum_wind_gust_direction_10_minutes,maximum_wind_gust_speed_10_minutes,maximum_wind_gust_direction_1_hour,maximum_wind_gust_speed_1_hour,maximum_wind_gust_direction_3_hours,maximum_wind_gust_speed_3_hours,rain_sensor_height,total_precipitation_1_hour,total_precipitation_3_hours,total_precipitation_6_hours,total_precipitation_12_hours,total_precipitation_24_hours\n0,20000,0,15015,15,15,1,2022,3,31,0,0,47.77706163,23.94046026,503,504.43,100940,10104,1448,5,298.15,294.55,80.4,3,1,1,0,0.004,10,-10,30,3,30,5,40,9,20,11,2,4.7,5.3,7.9,9.5,11.4", # noqa
+    "channel": "csv/test",
+    "notify": False,
+    "template": "aws_mappings.json"
+  }
+}
+
+expected_response = {
```
Comment (david-i-berry) [inline docker_compose_test/test_csv_publish.py:20]: de we want any of the _meta object returned?

File: wis2box_api/plugins/process/csv_publish.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            "title": "Notify",
+            "description": "Enable WIS2 notifications",
+            "schema": {"type": "boolean"},
+            "minOccurs": 1,
+            "maxOccurs": 1,
+            "metadata": None,
+            "default": True
+        }
+    },
+    'outputs': {
```
Comment (david-i-berry) [inline wis2box_api/plugins/process/csv_publish.py:84]: This appears to be different to that used in the tests, is this intended?

File: wis2box_api/plugins/process/csv_publish.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+from pathlib import Path
+
+import json
+
+LOGGER = logging.getLogger(__name__)
+
+PROCESS_METADATA = {
+    'version': '0.1.0',
+    'id': 'wis2box-csv-process',
```
[the remark this answers is not in the document]
Comment (maaikelimper) [reply wis2box_api/plugins/process/csv_publish.py:41]: OK good point let's agree on a naming convention, how about we call the file:
csv2bufr_process.py
And the process:
'id' : 'wis2box-csv2bufr-process'

and the same for synop2bufr ?

File: wis2box_api/plugins/process/csv_publish.py (hunk 2 above)
Comment (maaikelimper) [reply wis2box_api/plugins/process/csv_publish.py:84]: Yes I wanted one test returning raw data and another test returning file-urls , so I put notify=false to one of the tests.

File: wis2box_api/wis2box/station.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    def _load_stations(self):
+        """Load stations from WIS2 API
+
+        :returns: None
+        """
+
+        LOGGER.info("Loading stations from WIS2 API")
+
+        stations_url = "http://localhost/oapi/collections/stations/items"  # noqa
```
[the remark this answers is not in the document]
Comment (maaikelimper) [reply wis2box_api/wis2box/station.py:97]: I want to the process to take stations from the API itself and the name of the collections is hardcoded in wis2box already ... I see no use-case to let the user customize it ?

File: docker/pygeoapi-config.yml [hunk 5]
```diff
@@ -115,3 +115,19 @@ resources:
         type: process
         processor:
             name: wis2box_api.plugins.process.wmo_ra.WMORAProcessor
+
+    wis2box-bufr2geojson:
+      type: process
+      processor:
+        name: wis2box_api.plugins.process.bufr2geojson.Bufr2geojsonProcessor
+
+    wis2box-synop-process:
```
Comment (tomkralidis) [inline docker/pygeoapi-config.yml:124]: rename to `wis2box-synop2bufr`.

File: docker_compose_test/test_csv_publish.py
Comment (tomkralidis) [inline docker_compose_test/test_csv_publish.py:1]: Can these tests be put forth via pytest?

File: docker_compose_test/test_synop_publish.py
Comment (tomkralidis) [inline docker_compose_test/test_synop_publish.py:1]: Can these tests be put forth via pytest?

File: requirements.txt
Comment (tomkralidis) [inline requirements.txt:1]: Make alphabetical.

File: wis2box_api/plugins/process/bufr2geojson.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+###############################################################################
+import logging
```
Comment (tomkralidis) [inline wis2box_api/plugins/process/bufr2geojson.py:21]: Imports should be in the order of standard libary, 3rd party packages/modules, then local imports, separated by blank lines.  Example:

```python
import base64
import logging

import requests

from pygeoapi.process.base import BaseProcessor
```

File: wis2box_api/plugins/process/bufr2geojson.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+import base64
+
+from bufr2geojson import transform as as_geojson
+
+LOGGER = logging.getLogger(__name__)
+
+PROCESS_METADATA = {
+    'version': '0.1.0',
+    'id': 'wis2box-bufr2geojson',
+    'title': 'Convert BUFR to geoJSON',  # noqa
```
Comment (tomkralidis) [inline wis2box_api/plugins/process/bufr2geojson.py:33]: `s/geoJSON/GeoJSON/g`

File: wis2box_api/wis2box/handle.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+        Generate a checksum of message file
+        :param algorithm: secure hash algorithm (md5, sha512)
+        :returns: `tuple` of hexdigest and length
+        """
+
+        sh = getattr(hashlib, algorithm)()
+        sh.update(bytes)
+        return base64.b64encode(sh.digest()).decode()
+
+    def process_items(self, output_items: []):
```
Comment (tomkralidis) [inline wis2box_api/wis2box/handle.py:113]: Add function annotation.

File: wis2box_api/wis2box/minio.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+###############################################################################
+
```
Comment (tomkralidis) [inline wis2box_api/wis2box/minio.py:21]: Imports should be in the order of standard library, 3rd party packages/modules, then local imports, separated by blank lines.

File: wis2box_api/wis2box/minio.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        if self.source.startswith('https://'):
+            is_secure = True
+
+        self.client = Minio(endpoint=urlparsed.netloc,
+                            access_key=self.username,
+                            secret_key=self.password,
+                            secure=is_secure)
+
+    def get(self, identifier: str) -> Any:
```
Comment (tomkralidis) [inline wis2box_api/wis2box/minio.py:55]: add docstring.

File: wis2box_api/wis2box/minio.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+            data = response.data
+            response.close()
+            response.release_conn()
+        except Exception as err:
+            LOGGER.error(err)
+            raise err
+
+        return data
+
+    def put(self, data: bytes, identifier: str) -> bool:
```
Comment (tomkralidis) [inline wis2box_api/wis2box/minio.py:71]: add docstring.

File: wis2box_api/wis2box/minio.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        object_key = f'{self.channel}/{identifier}'
+        LOGGER.info(f'Putting data as object={object_key} in bucket={self.name}') # noqa
+        self.client.put_object(bucket_name=self.name,
+                               object_name=object_key,
+                               data=BytesIO(data), length=-1,
+                               part_size=10*1024*1024)
+
+        return True
+
+    def delete(self, identifier: str) -> bool:
```
Comment (tomkralidis) [inline wis2box_api/wis2box/minio.py:82]: add docstring.

File: wis2box_api/wis2box/station.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+###############################################################################
+import csv
+
+import io
```
Comment (tomkralidis) [inline wis2box_api/wis2box/station.py:23]: `from io import StringIO`, then `StringIO` in the caller below.

File: wis2box_api/wis2box/station.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomkralidis) [reply wis2box_api/wis2box/station.py:97]: +1 to derive from environment variable.  We need an `env.py` file in the same design as in `wis2box-management/wis2box/env.py`

File: wis2box_api/wis2box/storage.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+LOGGER = logging.getLogger(__name__)
+
+STORAGE_SOURCE = os.environ.get('WIS2BOX_STORAGE_SOURCE')
+STORAGE_USERNAME = os.environ.get('WIS2BOX_STORAGE_USERNAME')
+STORAGE_PASSWORD = os.environ.get('WIS2BOX_STORAGE_PASSWORD')
+
+
+class Storage():
+    """storage manager base class"""
+    def __init__(self, name, channel) -> None:
```
Comment (tomkralidis) [inline wis2box_api/wis2box/storage.py:37]: add/update docstring and function annotation.

File: wis2box_api/wis2box/storage.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+        self.name = name
+        self.channel = channel
+        self.source = STORAGE_SOURCE
+        self.username = STORAGE_USERNAME
+        self.password = STORAGE_PASSWORD
+
+    def get(self, identifier: str) -> Any:
```
Comment (tomkralidis) [inline wis2box_api/wis2box/storage.py:45]: add/update docstring and function annotation.

File: wis2box_api/wis2box/storage.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+        raise NotImplementedError
+
+    def put(self, data: bytes, identifier: str) -> bool:
```
Comment (tomkralidis) [inline wis2box_api/wis2box/storage.py:49]: add/update docstring and function annotation.

File: wis2box_api/wis2box/storage.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+        raise NotImplementedError
+
+    def delete(self, identifier: str) -> bool:
```
Comment (tomkralidis) [inline wis2box_api/wis2box/storage.py:53]: add/update docstring and function annotation.

File: wis2box_api/wis2box/storage.py [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+        raise NotImplementedError
```
Comment (tomkralidis) [inline wis2box_api/wis2box/storage.py:55]: `raise NotImplementedError()`

File: wis2box_api/wis2box/storage.py [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+        self.username = STORAGE_USERNAME
+        self.password = STORAGE_PASSWORD
+
+    def get(self, identifier: str) -> Any:
+
+        raise NotImplementedError
+
+    def put(self, data: bytes, identifier: str) -> bool:
+
+        raise NotImplementedError
```
Comment (tomkralidis) [inline wis2box_api/wis2box/storage.py:51]: `raise NotImplementedError()`

File: wis2box_api/wis2box/storage.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+        self.name = name
+        self.channel = channel
+        self.source = STORAGE_SOURCE
+        self.username = STORAGE_USERNAME
+        self.password = STORAGE_PASSWORD
+
+    def get(self, identifier: str) -> Any:
+
+        raise NotImplementedError
```
Comment (tomkralidis) [inline wis2box_api/wis2box/storage.py:47]: `raise NotImplementedError()`

**Review by tomkralidis — CHANGES_REQUESTED** [review]
Great work here!  See various change requests.  As well, for string quoting we should keep consistent between either single (preferred/consistent with other wis2box stack) or double quotes.

**Pull request merged**